test/opt/eliminate_dead_functions_test.cpp

```diff
@@ -0,0 +1,200 @@
+// limitations under the License.
+
+#include <string>
+#include <vector>
+
+#include <gmock/gmock.h>
+
+#include "assembly_builder.h"
+#include "pass_fixture.h"
+#include "pass_utils.h"
+
+
+template <typename T> std::vector<T> concat(const std::vector<T> &a, const std::vector<T> &b) {
+    std::vector<T> ret = std::vector<T>();
+    std::copy(a.begin(), a.end(), back_inserter(ret));
+    std::copy(b.begin(), b.end(), back_inserter(ret));
+    return ret;
+}
+
+namespace {
+
+using namespace spvtools;
+using ::testing::HasSubstr;
+
+using EliminateDeadFunctionsBasicTest = PassTest<::testing::Test>;
+
+TEST_F(EliminateDeadFunctionsBasicTest, BasicDeleteDeadFunction) {
+  // The function Dead should be removed because it is never called.
+  const std::vector<const char*> common_code = {
+      // clang-format off
+               "OpCapability Shader",
+          "%1 = OpExtInstImport \"GLSL.std.450\"",
+               "OpMemoryModel Logical GLSL450",
+               "OpEntryPoint Fragment %main \"main\" %gl_FragColor",
+               "OpExecutionMode %main OriginUpperLeft",
+               "OpSource GLSL 150",
+               "OpSourceExtension \"GL_GOOGLE_cpp_style_line_directive\"",
+               "OpSourceExtension \"GL_GOOGLE_include_directive\"",
+               "OpName %main \"main\"",
+               "OpName %Dead \"Dead\"",
+               "OpName %Constant \"Constant\"",
+               "OpName %gl_FragColor \"gl_FragColor\"",
+               "OpDecorate %gl_FragColor Location 0",
+       "%void = OpTypeVoid",
+          "%7 = OpTypeFunction %void",
+      "%float = OpTypeFloat 32",
+          "%9 = OpTypeFunction %float",
+  "%float_0_5 = OpConstant %float 0.5",
+  "%float_0_8 = OpConstant %float 0.8",
+    "%v4float = OpTypeVector %float 4",
+"%_ptr_Output_v4float = OpTypePointer Output %v4float",
+"%gl_FragColor = OpVariable %_ptr_Output_v4float Output",
+    "%float_1 = OpConstant %float 1",
+       "%main = OpFunction %void None %7",
+         "%15 = OpLabel",
+         "%16 = OpFunctionCall %float %Constant",
+         "%17 = OpFunctionCall %float %Constant",
+         "%18 = OpCompositeConstruct %v4float %16 %17 %float_0_8 %float_1",
+               "OpStore %gl_FragColor %18",
+               "OpReturn",
+               "OpFunctionEnd",
+  "%Constant = OpFunction %float None %9",
+         "%20 = OpLabel",
+               "OpReturnValue %float_0_8",
+               "OpFunctionEnd"
+      // clang-format on
+  };
+
+  const std::vector<const char*> dead_function = {
+      // clang-format off
+      "%Dead = OpFunction %float None %9",
+         "%19 = OpLabel",
+               "OpReturnValue %float_0_5",
+               "OpFunctionEnd",
+      // clang-format on
+  };
+
+  SetAssembleOptions(SPV_TEXT_TO_BINARY_OPTION_PRESERVE_NUMERIC_IDS);
+  SinglePassRunAndCheck<opt::EliminateDeadFunctionsPass>(
+      JoinAllInsts(concat(common_code, dead_function)),
+      JoinAllInsts(common_code), /* skip_nop = */ true);
```

Comment:
Worth noting that this also proves that appropriate OpNames have been removed.

Suggest extending the test case to include a decoration on an instruction inside one of the basic blocks.  For example, change the body to include a floating point add:

```      
     "%Dead = OpFunction %float None %9",
     "%19 = OpLabel",
     "%sum = OpFAdd %float %float_0_5 %float_0_5",
     "OpReturnValue %sum",
     "OpFunctionEnd",
```

And add a decoration in the decorations section:
 ```
OpDecorate %sum RelaxedPrecision
```

Then prove that even the decoration disappears.